pkg/hs/urbit-king/lib/Urbit/Vere/Dawn.hs

```diff
@@ -335,45 +335,72 @@ retrievePoint endpoint block ship =
       [x] -> pure x
       _   -> error "JSON server returned multiple return values."
 
-validateShipAndGetSponsor :: String -> TextBlockNum -> Seed -> RIO e Ship
-validateShipAndGetSponsor endpoint block (Seed ship life ring oaf) =
-  case clanFromShip ship of
-    Ob.Comet -> validateComet
-    Ob.Moon  -> validateMoon
-    _        -> validateRest
+validateFeedAndGetSponsor :: String
+                          -> TextBlockNum
+                          -> Feed
+                          -> RIO e (Seed, Ship)
+validateFeedAndGetSponsor endpoint block = \case
+    Feed0 s -> do
+      r <- validateSeed s
+      case r of
+        Left e  -> error e
+        Right r -> pure (s, r)
+    Feed1 s -> validateGerms s
+
   where
-    validateComet = do
-      -- A comet address is the fingerprint of the keypair
-      let shipFromPass = cometFingerprint $ ringToPass ring
-      when (ship /= shipFromPass) $
-        error ("comet name doesn't match fingerprint " <> show ship <> " vs " <>
-              show shipFromPass)
-      when (life /= 1) $
-        error ("comet can never be re-keyed")
-      pure (shipSein ship)
-
-    validateMoon = do
-      -- TODO: The current code in zuse does nothing, but we should be able to
-      -- try to validate the oath against the current as exists planet on
-      -- chain.
-      pure $ shipSein ship
-
-    validateRest = do
-      putStrLn ("boot: retrieving " <> renderShip ship <> "'s public keys")
-
-      whoP <- retrievePoint endpoint block ship
-      case epNet whoP of
-        Nothing -> error "ship not keyed"
-        Just (netLife, pass, contNum, (hasSponsor, who), _) -> do
-          when (netLife /= life) $
-              error ("keyfile life mismatch; keyfile claims life " <>
-                    show life <> ", but Azimuth claims life " <>
-                    show netLife)
-          when ((ringToPass ring) /= pass) $
-              error "keyfile does not match blockchain"
-          -- TODO: The hoon code does a breach check, but the C code never
-          -- supplies the data necessary for it to function.
-          pure who
+    validateGerms Germs{..} =
+      case gFeed of
+        []    -> error "no usable keys in keyfile"
+        (Germ{..}:f) -> do
+                  let seed = Seed gShip gLife gRing Nothing
+                  r :: Either String Ship
+                    <- validateSeed seed
+                  case r of
+                    Left _  -> validateGerms $ Germs gShip f
+                    Right r -> pure (seed, r)
+
+    validateSeed (Seed ship life ring oaf) =
+      case clanFromShip ship of
+        Ob.Comet -> validateComet
+        Ob.Moon  -> validateMoon
+        _        -> validateRest
+      where
+        validateComet = do
```

Comment:
so it seems that both validateComet and validateMoon are both pure, and only validateRest actually needs to be in the IO monad; do I have that right?

Reply:
I believe so, yes.

Reply:
I was being too passive. I think it's good practice to make things as pure as possible. The only reason it'd make sense to leave something that could be pure in IO is if you anticipate needing to add more IO operations to it later.